### PR TITLE
Refactor for Scraped

### PR DESCRIPTION
**Pull request by ondenman**
This is the first step towards scraping _all_ the member of the legislature. (Currently we're only picking up members with individual pages.)

## [Scraper Change checklist](https://github.com/everypolitician/everypolitician/wiki/Scraper-Change-checklist)
* [x] 1. scraper is on Morph.io under the "everypolitician-scrapers" group? (The scraper has been moved. https://morph.io/tmtmtmtm/croatia-parliament/ will need to be deleted.)
* [x] 2. scraper's GitHub "Website" link points at morph.io page?
* [x] 3. scraper is set to auto-run?
* [x] 4. scraper is archiving?
* [x] 5. legislature has a scraper webhook set? (The legislature has a Wikidata scraper but I am not able to check its settings as it is under

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -2,17 +2,19 @@
 # specified here will be installed and made available to your morph.io scraper.
 # Find out more: https://morph.io/documentation/ruby
 
-source "https://rubygems.org"
+source 'https://rubygems.org'
 git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
 
-ruby "2.0.0"
+ruby '2.3.1'
```
Comment (chrismytton) [inline Gemfile]: I think it probably makes sense to use the latest Ruby version, which is currently [2.3.3](https://www.ruby-lang.org/en/downloads/).

File: scraper.rb [hunk 2]
```diff
@@ -19,45 +21,15 @@ def noko_for(url)
...
-    # TODO: Chamges, e.g. http://www.sabor.hr/Default.aspx?sec=5358
-    term: term,
-    source: url.to_s,
-  }
-  data[:image] = URI.join(url, data[:image]).to_s unless data[:image].to_s.empty?
-
-  if data[:faction].to_s.empty?
-    data[:faction] = "Independent"
-    warn "No faction in #{data[:source]}: setting to #{data[:faction]}".red
+  members_page = MembersPage.new(response: Scraped::Request.new(url: url).response(decorators: [AbsoluteLinks]))
```
Comment (chrismytton) [inline scraper.rb]: This `members_page` variable doesn't seem to be used anywhere.

File: scraper.rb [hunk 3]
```diff
@@ -19,45 +21,15 @@ def noko_for(url)
...
+                    .members
+                    .map do |member_page|
+           member_page.to_h.merge(term: term)
+         end
+                    .each do |member_data|
```
Comment (chrismytton) [inline scraper.rb]: These lines are a bit hard to follow. I think it _might_ be better if you drop this `map` and instead merge the term info on the line below, when you call `ScraperWiki.save_sqlite`.

File: lib/members_page.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+require 'scraped'
+
+class MembersPage < Scraped::HTML
+  field :members do
+    noko.css('.liste2 .liste a').map do |a|
+      MemberPage.new(response: Scraped::Request.new(url: a.attr('href')).response)
```
Comment (chrismytton) [inline lib/members_page.rb]: [In other scrapers we've been using a `member_urls` method](https://github.com/search?l=&q=member_urls+user%3Aeverypolitician-scrapers&ref=advsearch&type=Code&utf8=%E2%9C%93) which just returns a list of the `href` attributes. These urls are then used in `scraper.rb` to create `Scraped::Request` and `MemberPage` instances.

The reasons for doing this are twofold. First if someone just wants to get a list of member urls then they have the overhead of every member page being fetched. Second it reduces the number of dependencies `MembersPage` has. Rather than depend on `MemberPage` and `Scraped::Request` it can just return an array of strings.

File: lib/member_page.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+  field :end_date do
+    noko.xpath('//td[b[contains(.,"End of parliamentary mandate:")]]/text()').text.split('/').reverse.join('-')
+  end
+
+  field :source do
+    url.to_s
+  end
+
+  # TODO: Chamges, e.g. http://www.sabor.hr/Default.aspx?sec=5358
```
Comment (chrismytton) [inline lib/member_page.rb]: Typo: Chamges

**Review by chrismytton — CHANGES_REQUESTED** [review]

**Comment by tmtmtmtm** [comment]
I've squashed all the commits down to make it easier to see what's going on and rebased over `master`.

File: lib/member_page.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  field :source do
+    url.to_s
+  end
+
+  # TODO: Changes, e.g. http://www.sabor.hr/Default.aspx?sec=5358
+
+  private
+
+  def dob_from(node)
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: As this is only called from a single place, and is tightly bound to the "Born on" text, there's no reason for this to be a pseudo-generic function that takes an argument — this can all just be a method tied to the date of birth.

File: scraper.rb [hunk 7]
```diff
@@ -20,45 +24,15 @@ def noko_for(url)
   Nokogiri::HTML(open(url).read)
 end
```
Comment (tmtmtmtm) [inline scraper.rb]: This method isn't used any more, so can be deleted.

File: scraper.rb [hunk 8]
```diff
@@ -20,45 +24,15 @@ def noko_for(url)
...
-    # TODO: Chamges, e.g. http://www.sabor.hr/Default.aspx?sec=5358
-    term:         term,
-    source:       url.to_s,
-  }
-  data[:image] = URI.join(url, data[:image]).to_s unless data[:image].to_s.empty?
-
-  if data[:faction].to_s.empty?
-    data[:faction] = 'Independent'
-    warn "No faction in #{data[:source]}: setting to #{data[:faction]}"
+  MembersPage.new(response: Scraped::Request.new(url: url).response(decorators: [AbsoluteLinks]))
```
Comment (tmtmtmtm) [inline scraper.rb]: This can now be updated in line with the modern approach to Decorators being on the class.

File: lib/decorators/absolute_links.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+require 'scraped'
+
+class AbsoluteLinks < Scraped::Response::Decorator
```
Comment (tmtmtmtm) [inline lib/decorators/absolute_links.rb]: This is now a built-in decorator, so you shouldn't need to include your own.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

**Comment by tmtmtmtm** [comment]
I've tidied this up a bit — back to @chrismytton for 👀

**Review by chrismytton — APPROVED** [review]
LGTM! ✨

**Comment by chrismytton** [comment]
This is fine from my POV, re-assigning to @tmtmtmtm as there's still a "requested changes" status on this PR.

**Review by tmtmtmtm — APPROVED** [review]
